### DISCUSSION TITLE
Keep alpha channel after padding image to square

### DISCUSSION THREAD

**Discussion by cp10111** [ImageMagick/ImageMagick#2361] [Help]
So I have this bit of code I copy and pasted. It pads an image with white pixels to make it square. That part works great. `convert /Users/tk01740/Desktop/93727_OC_C22833876_ALT4.tif -alpha on -virtual-pixel white -set option:distort:viewport \
     "%[fx:max(w,h)]x%[fx:max(w,h)]-%[fx:max((h-w)/2,0)]-%[fx:max((w-h)/2,0)]" \
     -filter point -distort SRT 0  +repage  /Users/tk01740/Desktop/93727_OC_C22833876_ALT4TESTED.tif
`
Problem is that it applies the alpha channel to the image destructively. I need the alpha channel to stay where it is and not be applied.
The photos attached are before/after. Padding works great. But I want to keep the alpha channel without applying it to the image. 
<img width="1014" alt="Screen Shot 2020-08-03 at 2 09 12 PM" src="https://user-images.githubusercontent.com/59060631/89218686-1a386100-d594-11ea-975c-def7c5d4b296.png">
alpha applied and gone
<img width="1128" alt="Screen Shot 2020-08-03 at 2 10 00 PM" src="https://user-images.githubusercontent.com/59060631/89218701-20c6d880-d594-11ea-90d4-0dc64d2e3bca.png">

Original Image: 
https://drive.google.com/file/d/1sPtD9zqEKdjZf-4m4JosMaDqY2o-9F0l/view?usp=sharing

**Comment by fmw42** [comment]
I do not think ImageMagick can write to the extra channels of a TIF or PSD file. I suspect any transparency is added as background transparency. Please post your original tiff file so we can test with it.

**Reply by cp10111** [reply]
Just added it to the post. https://drive.google.com/file/d/1sPtD9zqEKdjZf-4m4JosMaDqY2o-9F0l/view?usp=sharing

**Comment by fmw42** [comment]
This seems to work for me.

```
dims=`convert -quiet 93727_OC_C22833876_ALT4.tif -format "%[fx:max(w,h)]" info:`
convert -quiet 93727_OC_C22833876_ALT4.tif -gravity center -background none -extent ${dims}x${dims} -define tiff:alpha=unspecified x.tiff

```

<br>

If that works for you, then you can try adding -define tiff:alpha=unspecified to your command.

**Reply by cp10111** [reply]
That's closer than what I had before. The alpha channel stays where I need it, but It still seems to be applying the alpha to the image at some point. I'm losing pixels around the collar area and they don't come back when I try to fix it manually by editing the alpha in quickmask mode. 
<img width="1116" alt="Screen Shot 2020-08-03 at 3 36 10 PM" src="https://user-images.githubusercontent.com/59060631/89225062-38578e80-d59f-11ea-99da-ae24aa162b7f.png">

**Reply by fmw42** [reply]
I do not understand what pixels you are losing. The spots above the shoulders are in your original image. The only difference I see is that the background is black in the IM output and white in your original,  when alpha is off.

**Reply by cp10111** [reply]
On the left is the original image. Look around the tag. On the original, when I go into quick mask and paint the area around the tag white the original pixels show up. When I do the same on the IM processed image it just shows through to the black background. To me that says the alpha was applied to the image at some point. 
<img width="955" alt="Screen Shot 2020-08-03 at 4 01 44 PM" src="https://user-images.githubusercontent.com/59060631/89301611-7ac8ac00-d62f-11ea-96cc-fa6fd9e3dae4.png">

**Reply by fmw42** [reply]
One of the IM developers will need to comment further on this.

**Comment by snibgo** [comment]
IM reads 93727_OC_C22833876_ALT4.tif and sees one alpha channel. Many pixels are entirely transparent, with "correct" underlying colours.

Some IM operations will make any fully-transparent pixels transparent-black. This is an inevitable result from how alpha processing works: IM "loses" colour information for transparent pixels.

If you want to retain colour information for transparent pixels, you can extract alpha, then process colour and alpha separately, then "CopyOpacity" the alpha channel back. Windows BAT syntax:

```
magick ^
  -quiet ^
  93727_OC_C22833876_ALT4.tif ^
  -gravity center -background none ^
  ( -clone 0 ^
    -alpha Off ^
    -extent %%[fx:max(w,h)]x%%[fx:max(w,h)] ^
  ) ^
  ( -clone 0 ^
    -alpha Extract ^
    -extent %%[fx:max(w,h)]x%%[fx:max(w,h)] ^
  ) ^
  -delete 0 ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  -define tiff:alpha=unspecified ^
  x.tiff
```
Does that do what you want?

**Reply by cp10111** [reply]
Can I get it in Unix shell syntax? I don't have a windows machine to test on.

**Comment by fmw42** [comment]
Try

```
magick \
  -quiet \
  93727_OC_C22833876_ALT4.tif \
  -gravity center -background none \
  \( -clone 0 \
    -alpha Off \
    -extent "%[fx:max(w,h)]x%[fx:max(w,h)]" \
  \) \
  \( -clone 0 \
    -alpha Extract \
    -extent "%[fx:max(w,h)]x%[fx:max(w,h)]" \
  \) \
  -delete 0 \
  -alpha off \
  -compose CopyOpacity -composite \
  -define tiff:alpha=unspecified \
  x.tiff

```

**Reply by cp10111** [reply]
That is exactly what I need! Well done my friend!

**Comment by snibgo** [comment]
Silly me, I had forgotten about "-compose Copy". This simplifies the solution:

```
%IMG7%magick ^
  -quiet ^
  93727_OC_C22833876_ALT4.tif ^
  -gravity center -background None ^
  -compose Copy ^
  -extent %%[fx:max(w,h)]x%%[fx:max(w,h)] ^
  -define tiff:alpha=unspecified ^
  x.tiff
```
